corehq/util/pagination.py

```diff
@@ -0,0 +1,272 @@
+import hashlib
+from datetime import datetime
+
+from couchdbkit import ResourceNotFound
+from jsonobject.properties import ListProperty, BooleanProperty
+
+from dimagi.ext.jsonobject import JsonObject, StringProperty, DateTimeProperty, DictProperty
+from dimagi.utils.couch.database import get_db
+
+
+class PaginationEventHandler(object):
+    def page_start(self, total_emitted, *args, **kwargs):
+        """Called prior to getting each page of data
+
+        :param total_emitted: total count of data items yielded
+        :param args: Argument list that was passed to the ``data_function`` for this page
+        :param kwargs: Keyword arguments that were passed to the ``data_function`` for this page
+        """
+        pass
+
+    def page_end(self, total_emitted, duration, *args, **kwargs):
+        """Called at the end of each page of data
+
+        :param total_emitted: total count of data items yielded
+        :param duration: the duration in ms that it took to yield this page
+        :param args: Argument list that was passed to the ``data_function`` for this page
+        :param kwargs: Keyword arguments that were passed to the ``data_function`` for this page
+        """
+        pass
+
+
+class DelegatingPaginationEventHandler(PaginationEventHandler):
+    def __init__(self, handlers=None):
+        self.handlers = handlers or []
+
+    def add_handler(self, handler):
+        self.handlers.append(handler)
+
+    def page_start(self, total_emitted, *args, **kwargs):
+        for handler in self.handlers:
+            handler.page_start(total_emitted, *args, **kwargs)
+
+    def page_end(self, total_emitted, duration, *args, **kwargs):
+        for handler in self.handlers:
+            handler.page_end(total_emitted, duration, *args, **kwargs)
+
+
+class ArgsProvider(object):
+    def get_initial_args(self):
+        """Return the initial set of args and kwargs
+
+        :returns: tuple of args list and kwargs dict"""
+        raise NotImplementedError
+
+    def get_next_args(self, last_item, *last_args, **last_kwargs):
+        """Return the next set of args and kwargs
+
+        :returns: tuple of args list and kwargs dict
+        :raises: StopIteration to end the iteration
+        """
+        raise StopIteration
+
+
+def paginate_function(data_function, args_provider, event_handler=PaginationEventHandler()):
+    """
+    Repeatedly call a data provider function with successive sets of arguments provided
+    by the ``args_provider``
+
+    :param data_function: function to paginate. Must return an list of data elements.
+    :param args_provider: An instance of the ``ArgsProvider`` class.
```

Comment:
Might want to say something about what this is used for (provides args for `data_function`)